### PR TITLE
Fix dh parameter labels and add method for calculating covariance in tangent space

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
This addresses two things and exposes one issue ~~and not sure how to fix~~.
- The parameter labels in stored in the optimize function were not correct because Eigen is column major so the lables need to be stored the same way.
- The compute covariance includes all parameters even if a parameter block is constant or sub parameters are constant. This adds another function but was thinking about updating the existing to do this. @marip8 Let me know what you think about updating the existing functions to do this instead of having a separate function?
- Also noticed something odd where on some tests the parameter block was in an incorrect order. I have added a check and now the unit tests are failing because this is being detected. Some how the target dh parameter block and camera dh parameter block is being swapped. @marip8 Do you know why this would be happening.

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
It turns out that some the code assumes the vector of parameter blocks returned from ceres would be in the same order added. This is not the case because ceres is store it in a map and where requested it iterates over the map and create a vector using the key so the order may not be the same.

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 1]
```diff
@@ -171,37 +171,40 @@ CovarianceResult computeCovariance(ceres::Problem &problem,
...
-    int block_size = problem.ParameterBlockSize(parameter_blocks[index_block]);
-    if (static_cast<std::size_t>(block_size) != parameter_names[index_block].size())
+    int block_size = problem.ParameterBlockSize(b);
+    if (static_cast<std::size_t>(block_size) != param_names.at(b).size())
     {
       std::stringstream ss;
-      ss << "Number of parameter labels provided for block " << index_block << " does not match actual number of parameters in that block: " \
-         << "have " << parameter_names[index_block].size() << " labels and " << block_size << " parameters";
+      ss << "Number of parameter labels provided for block does not match actual number of parameters in that block: " \
+         << "have " << param_names.at(b).size() << " labels and " << block_size << " parameters";
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:199]: Add the name of the block in the error message

```suggestion
      ss << "Number of parameter labels provided for block '" << param_names.at(b).first << "' does not match actual number of parameters in that block: " \
         << "have " << param_names.at(b).size() << " labels and " << block_size << " parameters";
```

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 2]
```diff
@@ -169,4 +169,118 @@ CovarianceResult computeCovariance(ceres::Problem &problem,
 
   return res;
 }
+
+CovarianceResult computeCovarianceInTangentSpace(ceres::Problem &problem,
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:173]: Did you look at using [this function in Ceres](http://ceres-solver.org/nnls_covariance.html#_CPPv4NK5ceres32GetCovarianceBlockInTangentSpaceEPKdPKdPd) to get the covariance block of the tangent space? Seems like it might be equivalent and much easier than adding a whole new function

**Review by marip8 — CHANGES_REQUESTED** [review]
Overall I'm on board with this change. I think it's worth looking into a simpler way extracting the covariance of the tangent space though

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp (hunk 2 above)
Comment (Levi-Armstrong) [reply rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:173]: I tried that first but was unsuccessful due to runtime crashing. I assumed it was because the size of the covariance matrix being passed in was not correct, which led me to to believe it was including other parameters that I wanted to ignore so I ended up with this.

**Comment by Levi-Armstrong** [comment]
@marip8 Do you think all the function should be updated to return this as the covariance or keep it as a separate?

**Comment by Levi-Armstrong** [comment]
@marip8 you good with me merging this?

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 3]
```diff
@@ -116,57 +169,81 @@ CovarianceResult computeCovariance(ceres::Problem &problem,
...
 
-  // 5. Create NamedParams for covariance and correlation results, which include labels and values for the parameters. Uses top-right triangular part of matrix.
-  Eigen::Index col_start = 0;
-  for (Eigen::Index row = 0; row < correlation_matrix.rows(); row++)
+  if (parameter_blocks.size() != param_masks.size())
+    throw CovarianceException("Provided vector param_masks is not same length as provided number of parameter blocks");
+
+  Eigen::Index n_params_in_selected = 0;
+  std::vector<Eigen::Index> tagent_space_indices;
+  std::vector<std::string> tagent_space_labels;
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:194]:
```suggestion
  std::vector<Eigen::Index> tangent_space_indices;
  std::vector<std::string> tangent_space_labels;
```

File: rct_optimizations/src/rct_optimizations/dh_chain_kinematic_calibration.cpp [hunk 4]
```diff
@@ -416,6 +480,9 @@ KinematicCalibrationResult optimize(const KinematicCalibrationProblemPose6D &par
     problem.AddResidualBlock(cost_block, nullptr, target_chain_dh_offsets.data());
   }
 
+  // Print optimization parameter labels
+  printOptimizationLabels(problem, param_names, param_labels, param_masks);
+
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/dh_chain_kinematic_calibration.cpp:485]: Is this really needed or is this for debug?

**Review by marip8 — CHANGES_REQUESTED** [review]
I haven't had a chance to fully look through this yet; I'll look at it in more detail tonight. One thought though: does it ever make sense not to compute the covariance in the tangent space? It seems like covariance between static variables is not valid

**Comment by Levi-Armstrong** [comment]
> I haven't had a chance to fully look through this yet; I'll look at it in more detail tonight. One thought though: does it ever make sense not to compute the covariance in the tangent space? It seems like covariance between static variables is not valid

Yea, that was my thought is it is not needed. If you wanted the full covariance you would just pass it an empty masks. If you onboard I can just remove the new function and update the documentation on the other functions. Let me know if you agree and I will make the change.

File: rct_optimizations/src/rct_optimizations/dh_chain_kinematic_calibration.cpp (hunk 4 above)
Comment (Levi-Armstrong) [reply rct_optimizations/src/rct_optimizations/dh_chain_kinematic_calibration.cpp:485]: Just for debug.

**Comment by marip8** [comment]
> If you onboard I can just remove the new function and update the documentation on the other functions. Let me know if you agree and I will make the change.

I think that makes sense; I'm on board with doing that. Like you mentioned, if you want the full covariance you can just pass empty masks (or we could make that the default value for that argument)

File: rct_common/include/rct_common/extrinsic_data_set.h [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#ifndef EXTRINSIC_DATA_SET_H
+#define EXTRINSIC_DATA_SET_H
+
+#endif // EXTRINSIC_DATA_SET_H
```
Comment (marip8) [inline rct_common/include/rct_common/extrinsic_data_set.h:4]: Is this necessary?

Comment (Levi-Armstrong) [reply rct_common/include/rct_common/extrinsic_data_set.h:4]: Nope

File: rct_optimizations/include/rct_optimizations/covariance_analysis.h [hunk 6]
```diff
@@ -30,40 +38,48 @@ Eigen::MatrixXd computeCorrelationsFromCovariance(const Eigen::MatrixXd& covaria
...
+                                   const std::map<const double *, std::vector<int> >& param_masks = std::map<const double *, std::vector<int>>(),
                                    const ceres::Covariance::Options& options = DefaultCovarianceOptions());
 
 /**
  * @brief Compute covariance results for specified parameter blocks in a Ceres optimization problem and label them with the provided names.
+ * @details This excludes all parameters set to be constant during the solve
  * @param problem The Ceres problem (after optimization).
  * @param parameter_blocks Specific parameter blocks for which covariance will be calculated.
  * @param parameter_names Labels for optimization parameters in the specified blocks.
+ * @param param_masks The mask parameters associated with parameter block
```
Comment (marip8) [inline rct_optimizations/include/rct_optimizations/covariance_analysis.h:82]:
```suggestion
 * @param param_masks Map of the parameter block pointer and the indices of the parameters within that block to be excluded from the covariance calculation
```

File: rct_optimizations/include/rct_optimizations/covariance_analysis.h [hunk 7]
```diff
@@ -73,6 +89,8 @@ CovarianceResult computeCovariance(ceres::Problem &problem,
  */
 CovarianceResult computeCovariance(ceres::Problem &problem,
                                    const std::vector<const double *>& parameter_blocks,
-                                   const std::vector<std::vector<std::string>>& parameter_names,
+                                   const std::map<const double *, std::vector<std::string> > &param_names,
+                                   const std::map<const double *, std::vector<int> >& param_masks = std::map<const double *, std::vector<int>>(),
```
Comment (marip8) [inline rct_optimizations/include/rct_optimizations/covariance_analysis.h:93]:
```suggestion
                                   const std::map<const double*, std::vector<std::string>> &param_names,
                                   const std::map<const double*, std::vector<int>>& param_masks = std::map<const double*, std::vector<int>>(),
```

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 8]
```diff
@@ -36,22 +36,73 @@ Eigen::MatrixXd computeCorrelationsFromCovariance(const Eigen::MatrixXd& covaria
...
+      res.covariances.push_back(p_cov);
+    }
+    col_start++;
+  }
+
+  return res;
+}
+
 CovarianceResult computeCovariance(ceres::Problem &problem,
+                                   const std::map<const double *, std::vector<int> > &param_masks,
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:89]:
```suggestion
                                   const std::map<const double*, std::vector<int>> &param_masks,
```

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 9]
```diff
@@ -36,22 +36,73 @@ Eigen::MatrixXd computeCorrelationsFromCovariance(const Eigen::MatrixXd& covaria
...
+  return computeCovariance(problem, blocks_const, param_masks, options);
 }
 
 CovarianceResult computeCovariance(ceres::Problem &problem,
                                    const std::vector<const double *>& parameter_blocks,
+                                   const std::map<const double *, std::vector<int> > &param_masks,
                                    const ceres::Covariance::Options& options)
 {
-  std::vector<std::vector<std::string>> dummy_param_names;
+  std::map<const double *, std::vector<std::string>> dummy_param_names;
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:105]:
```suggestion
  std::map<const double*, std::vector<std::string>> dummy_param_names;
```

File: rct_optimizations/src/rct_optimizations/covariance_analysis.cpp [hunk 10]
```diff
@@ -61,15 +112,16 @@ CovarianceResult computeCovariance(ceres::Problem &problem,
...
 
-  return computeCovariance(problem, parameter_blocks, dummy_param_names, options);
+  return computeCovariance(problem, parameter_blocks, dummy_param_names, param_masks, options);
 
 }
 
 CovarianceResult computeCovariance(ceres::Problem &problem,
-                                   const std::vector<std::vector<std::string>>& parameter_names,
+                                   const std::map<const double *, std::vector<std::string> > &param_names,
+                                   const std::map<const double *, std::vector<int> > &param_masks,
```
Comment (marip8) [inline rct_optimizations/src/rct_optimizations/covariance_analysis.cpp:124]:
```suggestion
                                   const std::map<const double*, std::vector<std::string>> &param_names,
                                   const std::map<const double*, std::vector<int>> &param_masks,
```

File: rct_ros_tools/include/rct_ros_tools/print_utils.h [hunk 11]
```diff
@@ -1,144 +1,21 @@
...
+using rct_common::getStringQuaternion;
+using rct_common::getStringRPY;
+using rct_common::getStringXYZ;
+using rct_common::printCameraDistortion;
+using rct_common::printCameraIntrinsics;
+using rct_common::printNewLine;
+using rct_common::printOptResults;
+using rct_common::printTitle;
+using rct_common::printTransform;
+using rct_common::printTransformDiff;
```
Comment (marip8) [inline rct_ros_tools/include/rct_ros_tools/print_utils.h:19]: This seems unnecessary; let's get rid of this file and just change references to this function to use the `rct_common` namespace

Comment (Levi-Armstrong) [reply rct_ros_tools/include/rct_ros_tools/print_utils.h:19]: This was to prevent breaking changes, but I am fine without removing.

File: rct_optimizations/include/rct_optimizations/covariance_analysis.h [hunk 12]
```diff
@@ -30,40 +38,48 @@ Eigen::MatrixXd computeCorrelationsFromCovariance(const Eigen::MatrixXd& covaria
...
  */
 CovarianceResult computeCovariance(ceres::Problem &problem,
-                                   const std::vector<std::vector<std::string>>& parameter_names,
+                                   const std::map<const double *, std::vector<std::string> > &param_names,
+                                   const std::map<const double *, std::vector<int> >& param_masks = std::map<const double *, std::vector<int>>(),
                                    const ceres::Covariance::Options& options = DefaultCovarianceOptions());
 
 /**
  * @brief Compute covariance results for specified parameter blocks in a Ceres optimization problem and label them with the provided names.
+ * @details This excludes all parameters set to be constant during the solve
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply rct_optimizations/include/rct_optimizations/covariance_analysis.h:78]: This is a valid comment. This does occur regardless of the mask now.

**Comment by Levi-Armstrong** [comment]
> It seems like this PR is making 5 sets of unrelated changes:
> 
> * Removing the Xenial build
> * Fix for YAML include
> * Moving print headers to `rct_common`
> * Adding a helper to the DH chain to get a relative transform
> * Updating the covariance calculation
> 
> I'm okay with making these changes, but I would really prefer if we merged them as separate PRs in the order above. It makes it hard to look back and understand where changes were introduced if too many unrelated ones are made in the same PR

I wound not agree these are unrelated changes. These were all necessary to get something that was broken tested and fixed. I will clean up the commit history but I am not going to separate them into different PRs. In the end the only thing that is important is that the commit history reflect these changes.

**Comment by Levi-Armstrong** [comment]
I will follow up with a different PR to add a clang format file, CI Job and format the entire repo.

File: rct_optimizations/include/rct_optimizations/covariance_analysis.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (marip8) [reply rct_optimizations/include/rct_optimizations/covariance_analysis.h:81]: Okay; I assume this is because we are checking `isParameterBlockConstant` in the function?

**Comment by marip8** [comment]
> I will clean up the commit history but I am not going to separate them into different PRs. In the end the only thing that is important is that the commit history reflect these changes.

That's fine; the main thing is being able to see the individual changes

**Comment by Levi-Armstrong** [comment]
> > I will clean up the commit history but I am not going to separate them into different PRs. In the end the only thing that is important is that the commit history reflect these changes.
> 
> That's fine; the main thing is being able to see the individual changes

Sorry, I reread my comment. Please don't take it the wrong way. I understand your point but in this case I needed all of these changes with the exception of the moving the print utils in one branch to for everything to work on a machine. The CI change was also need because some change I made broke the xenial build and I did not want to spend time fixing it for something that was EOL.

File: rct_ros_tools/include/rct_ros_tools/print_utils.h (hunk 11 above)
Comment (Levi-Armstrong) [reply rct_ros_tools/include/rct_ros_tools/print_utils.h]: I removed this header and updated everywhere it was used. I add these changes the original commit so they are all in one commit.

**Comment by Levi-Armstrong** [comment]
I think everything is address except the formatting issues, which I will fix in a followup PR using clang format.

File: rct_common/include/rct_common/extrinsic_data_set.h (hunk 5 above)
Comment (marip8) [reply rct_common/include/rct_common/extrinsic_data_set.h:4]: Let's delete this file

**Comment by Levi-Armstrong** [comment]
@marip8 I created a PR add the linters which I will rebase and run after this is merged.

**Comment by marip8** [comment]
Can you remove the empty extrinsic data header file? It still looks to be added in this PR. Otherwise I think it should be ready to merge

**Comment by Levi-Armstrong** [comment]
@marip8 I removed that file.

File: rct_optimizations/include/rct_optimizations/covariance_analysis.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply rct_optimizations/include/rct_optimizations/covariance_analysis.h:81]: Yea, not sure if there is a better way to handle this. When creating the problem we use the mask data and if all parameters in a group are masked it sets it to constant. Though I could possible remove this check and solely rely on the mask data. @marip8 What do you thing?

[the remark this answers is not in the document]
Comment (marip8) [reply rct_optimizations/include/rct_optimizations/covariance_analysis.h:81]: I'm inclined to say that we remove this check and let the user explicitly provide the mask. The covariance between a variable and a constant is 0, so worst case scenario is that you'll see a bunch of zeros if you forget to mask a constant block. 

Following the idea of autmatically removing constant parameters from the covariance: another thing that makes this tricky is that individual parameters can be held constant using a `SubsetParameterization`. We could also automatically remove those from the covariance matrix (without specifiying a mask) by calling `GetCovarianceMatrixInTangentSpace` instead of `GetCovarianceMatrix`. At this point however, the size/indices of your covariance matrix no longer correspond to the parameter block labels and indices. This would also be made worse if any other local parameterizations were used (e.g., the local parameterization that represents quaternions with 3 values instead of 4).

TLDR: let's get the full covariance matrix (like we're already doing), not automatically remove constant blocks, and rely explicitly on the masks to show the user the covariances they asked to see

[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply rct_optimizations/include/rct_optimizations/covariance_analysis.h]: I just added a commit to only use the mask info.

**Review by marip8 — APPROVED** [review]

**Comment by Levi-Armstrong** [comment]
@marip8 Thanks for reviewing this!
